### PR TITLE
Add Conservative from Primitive for Newtonian Euler

### PR DESCRIPTION
**Pull request by carmaza**
### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files) or `tests/Unit/TestingFramework.hpp` (only in tests)
  2. Blank line (only in cpp files and tests)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### CODE REVIEW

File: tests/Unit/Evolution/Systems/NewtonianEuler/Test_ConservativeFromPrimitive.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+}  // namespace
+
+SPECTRE_TEST_CASE(
+    "Unit.Evolution.Systems.NewtonianEuler.ConservativeFromPrimitive",
+    "[Unit][Evolution]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "Evolution/Systems/NewtonianEuler"};
+
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/NewtonianEuler/Test_ConservativeFromPrimitive.cpp:30]: In CheckWithRandomValues.hpp add 3 macros
```
# define CHECK_FOR_DATAVECTORS(...) \
const DataVector dv(5); \
__VA_ARGS__<1>(dv); \
...
```
and also CHECK_FOR_DOUBLES and CHECK_FOR_DOUBLES_AND_DATAVECTORS

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@nilsdeppe please take a look

**Review by wthrowe — DISMISSED** [review]

File: tests/Unit/Pypp/CheckWithRandomValues.hpp [hunk 2]
```diff
@@ -6,25 +6,37 @@
...
 #include <boost/preprocessor/tuple/push_back.hpp>
 #include <boost/preprocessor/tuple/push_front.hpp>
+#include <boost/preprocessor/tuple/rem.hpp>
 #include <boost/preprocessor/tuple/size.hpp>
+#include <boost/preprocessor/tuple/to_array.hpp>
 #include <boost/preprocessor/tuple/to_list.hpp>
 #include <boost/preprocessor/variadic/elem.hpp>
 #include <boost/preprocessor/variadic/to_list.hpp>
 #include <boost/preprocessor/variadic/to_tuple.hpp>
+#include <boost/vmd/detail/is_empty.hpp>
```
Comment (nilsdeppe) [inline tests/Unit/Pypp/CheckWithRandomValues.hpp]: For some of these, specifically the `detail` ones, you should modify the IWYU map file in `tools/Iwyu/boost-all.imp` (just append to the end). For example,
```
{ include: ["<boost/vmd/detail/is_empty.hpp>", private,
                "<boost/vmd/is_empty.hpp>", public]},
```

File: tests/Unit/Evolution/Systems/NewtonianEuler/Test_ConservativeFromPrimitive.cpp [hunk 3]
```diff
@@ -29,13 +28,6 @@ SPECTRE_TEST_CASE(
...
-  const double d = std::numeric_limits<double>::signaling_NaN();
-  test_conservative_from_primitive<1>(d);
-  test_conservative_from_primitive<2>(d);
-  test_conservative_from_primitive<3>(d);
-
-  const DataVector dv(5);
-  test_conservative_from_primitive<1>(dv);
-  test_conservative_from_primitive<2>(dv);
-  test_conservative_from_primitive<3>(dv);
+  GENERATE_UNINITIALIZED_DOUBLE_AND_DATAVECTOR;
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/NewtonianEuler/Test_ConservativeFromPrimitive.cpp]: Just move this into the previous commit :)

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm happy for you to squash immediately :)

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@nilsdeppe please look at

File: tests/Unit/Pypp/CheckWithRandomValues.hpp (hunk 2 above)
Comment (nilsdeppe) [reply tests/Unit/Pypp/CheckWithRandomValues.hpp]: ping ping ping

**Review by nilsdeppe — DISMISSED** [review]

**Review by mar-celine — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Comment by nilsdeppe** [comment]
No, get this working.

**Comment by nilsdeppe** [comment]
It's a matter of reading, come on, it's not hard.

File: tests/Unit/Pypp/CheckWithRandomValues.hpp [hunk 4]
```diff
@@ -6,21 +6,32 @@
 #include "tests/Unit/TestingFramework.hpp"
 
 // IWYU pragma: begin_exports
+#include <boost/preprocessor/arithmetic/inc.hpp>
 #include <boost/preprocessor/comparison/equal.hpp>
+#include <boost/preprocessor/comparison/not_equal.hpp>
+#include <boost/preprocessor/control/expr_iif.hpp>
 #include <boost/preprocessor/control/if.hpp>
 #include <boost/preprocessor/debug/assert.hpp>
+#include <boost/preprocessor/detail/check.hpp>
```
Comment (nilsdeppe) [inline tests/Unit/Pypp/CheckWithRandomValues.hpp:15]: - `preprocessor/list/adt.hpp`
- `preprocessor/facilities/apply.hpp`
- `parameter/preprocessor.hpp`

Are the interfaces. I did:
```
$ rg 'check.hpp'
detail/is_binary.hpp
16:# include <boost/preprocessor/detail/check.hpp>

detail/is_nullary.hpp
16:# include <boost/preprocessor/detail/check.hpp>

detail/is_unary.hpp
16:# include <boost/preprocessor/detail/check.hpp>
$ rg 'is_binary.hpp'
list/adt.hpp
18:# include <boost/preprocessor/detail/is_binary.hpp>
$ rg 'is_unary.hpp'
facilities/apply.hpp
17:# include <boost/preprocessor/detail/is_unary.hpp>
$ rg 'detail/is_nullary.hpp'
parameter/preprocessor.hpp
35:# include <boost/preprocessor/detail/is_nullary.hpp>
```

**Review by nilsdeppe — COMMENTED** [review]
I'm requesting this change...

**Pull request merged**